src/lib/Live/Eval.hs

```diff
@@ -56,77 +97,207 @@ data CacheState = CacheState
+  remaining'' <- mapM makeNewBlockId remaining'
+  updateResultList $ map getNodeId $ evaluated ++ reused ++ remaining''
+  putPending $ reused ++ remaining''
+  maybeLaunchWorker
+
+-- See which blocks already have completed values and reuse those.
+tryEvalBlocksCached :: Driver m
+                    => [SourceBlock]
+                    -> m ([WithId SourceBlock], [SourceBlock])
 tryEvalBlocksCached [] = return ([], [])
 tryEvalBlocksCached blocks@(block:rest) = do
-  (env, cache) <- get
-  case M.lookup (block, env) (evalCache cache) of
+  env <- getTopState
+  res <- lookupCache block env
+  case res of
     Nothing -> return ([], blocks)
     Just (blockId, env') -> do
       let block' = WithId blockId block
-      updateTopState env'
+      putTopState env'
       (evaluated, remaining) <- tryEvalBlocksCached rest
       return (block':evaluated, remaining)
 
-evalBlock :: WithId SourceBlock -> DriverM ()
-evalBlock (WithId blockId block) = do
-  oldState <- gets fst
-  opts <- asks fst
-  (result, s) <- liftIO $ evalSourceBlockIO opts (withoutId oldState) block
-  resultsChan <- asks snd
+-- See whether the formerly active worker (if any) is still doing
+-- something useful given the list of blocks we are waiting to finish;
+-- if so reuse it, and if not kill it.
+tryReuseWorker :: Driver m
+               => WithId TopStateEx
+               -> WorkerSpec
+               -> [SourceBlock]
+               -> m ([WithId SourceBlock], [SourceBlock])
+tryReuseWorker _ w [] =
+  liftIO (forM_ w (killThread . fst)) >> return ([], [])
+tryReuseWorker _ Nothing blocks =
+  return ([], blocks)
+tryReuseWorker oldEnv w@(Just (_, oldNext)) (next:rest) = do
+  curEnv <- getTopState
+  if (curEnv == oldEnv) && (withoutId oldNext == next) then do
+    -- Reuse the worker
+    putWorker w
+    return ([oldNext], rest)
+  else
+    liftIO (forM_ w (killThread . fst)) >> return ([], next:rest)
+
+processWork :: Driver m
+            => WithId TopStateEx
+            -> WithId SourceBlock
+            -> (Result, TopStateEx)
+            -> m ()
+processWork oldState block answer = do
+  -- The computed result is true regardless of whether this is the
+  -- worker we are waiting for or not, and therefore safe to cache
+  -- outside the `when` clause.  There is a narrow benefit here: if a
+  -- worker completes normally while we're processing a FileChanged
+  -- message, it can send a sound WorkComplete message before we
+  -- actually kill it.  We record that result in case the user edits
+  -- back to a state where it can be shown.
+  newState <- recordTruth oldState block answer
+  curState <- getTopState
+  waitingFor <- getWorkingBlock
+  when (oldState == curState
+        && (fmap withoutId waitingFor == Just (withoutId block))) $ do
+    -- We only update our working state if this message is, in fact,
+    -- from the worker we are currently waiting for.
+    rotateWorkingState newState
```

Comment:
I feel like there is an easier way to determine whether the message comes from our current worker: just ask the worker to send back a unique integer once it's done, and remember the integer the current worker is supposed to send you back. You might still get those out-of-band messages, but the check is pretty trivial then.

I think that would be an improvement, but I don't feel strongly enough to block this PR on it. It's still an improvement. So I'll leave the decision up to you.

Reply:
Yes, that would probably work, but I was going for a more semantic check here.